### PR TITLE
Use shared egress ACL table for PFCWD in BRCM DNX platform

### PR DESCRIPTION
**Pull request by ysmanman**
Today PFCWD uses separate egress ACL tables, one table for each traffic class. This may result in ACL resource overflow when other features that also need egress ACL table is enabled at the same time, e.g., MACSEC. ( See https://github.com/sonic-net/sonic-buildimage/issues/17839 for details).

The fix is to use shared egress ACL table for PFCWD in BRCM DNX platform.

**What I did**

**Why I did it**

**How I verified it**

**Details if related**

### CODE REVIEW

**Comment by ysmanman** [comment]
Add @kenneth-arista @arlakshm @vmittal-msft for viz.

**Comment by arlakshm** [comment]
Please add UT.

**Comment by arlakshm** [comment]
Sai changes still now complete. These are sonic changes for this functionality change. PR to be merged after the SAI changes are integrated.

**Comment by arlakshm** [comment]
@saksarav-nokia

**Review by vmittal-msft — APPROVED** [review]
Changes looks fine. 
1. Can you please share all the test run results to verify this functionality? Manual, sonic-mgmt, ixia tests etc
2. Please share syslog message change due to this. current/new both.
3. Please test with/without mac sec enabled ports.
4.  Also, please post results from pizzabox.
5. Are these swss changes complete or expecting more after applying second patch from BRCM ?
6. Please make sure to verify pfcwd drop counters to make sure they are still reporting drops.

Thanks
Vineet

**Comment by vmittal-msft** [comment]
Also, please add test cases to cover changes.

**Comment by ysmanman** [comment]
The test added in this PR depends on https://github.com/sonic-net/sonic-buildimage/pull/19108

**Comment by ysmanman** [comment]
> Changes looks fine.
> 
> 1. Can you please share all the test run results to verify this functionality? Manual, sonic-mgmt, ixia tests etc
> 2. Please share syslog message change due to this. current/new both.
> 3. Please test with/without mac sec enabled ports.
> 4. Also, please post results from pizzabox.
> 5. Are these swss changes complete or expecting more after applying second patch from BRCM ?
> 6. Please make sure to verify pfcwd drop counters to make sure they are still reporting drops.
> 
> Thanks Vineet



> Changes looks fine.
> 
> 1. Can you please share all the test run results to verify this functionality? Manual, sonic-mgmt, ixia tests etc

We completed the testing with ixia, and shared egress acl worked fine without seeing any issue.

> 2. Please share syslog message change due to this. current/new both.

The change adds the following new syslog when binging eress ACL table to switch:
```
May 22 17:34:39.194760 cmp329-6 NOTICE swss0#orchagent: :- bindEgrAclTableToSwitch: Bind egress acl table EgressTable_PfcWdAclHandler_Drop to switch
```

> 3. Please test with/without mac sec enabled ports.

Yes, we tested PFCWD alone and also PFCWD + MACsec.

> 4. Also, please post results from pizzabox.

We will provide the result when the testing is done.

> 5. Are these swss changes complete or expecting more after applying second patch from BRCM ?

The swss change is complete.

> 6. Please make sure to verify pfcwd drop counters to make sure they are still reporting drops.

Yes,  the drop counters worked fine.

> 
> Thanks Vineet

**Comment by ysmanman** [comment]
> Also, please add test cases to cover changes.



> Also, please add test cases to cover changes.

Hi @vmittal-msft, test was added.

**Comment by ysmanman** [comment]
Hi @vmittal-msft , the test in this PR depends on https://github.com/sonic-net/sonic-buildimage/pull/19108.

**Comment by vmittal-msft** [comment]
@ysmanman please rebase and check if PR checks still failing.

**Comment by ysmanman** [comment]
> @ysmanman please rebase and check if PR checks still failing.

@vmittal-msft Rebased just now. Please note the vs test added in this PR cannot pass without https://github.com/sonic-net/sonic-buildimage/pull/19108.

**Comment by vmittal-msft** [comment]
@ysmanman  Can you please fix compilation errors ?

**Comment by ysmanman** [comment]
> @ysmanman Can you please fix compilation errors ?

Hi @vmittal-msft , compilation error was fixed. vs test failed. The test requires https://github.com/sonic-net/sonic-buildimage/pull/19108, which was merged recently. Will sonic-swss build pick up the recently merged sonic-buildimage change automatically?

**Comment by vmittal-msft** [comment]
I am not sure if it does.  @prsunny Can you please help? Fix for VS test failure is part of sonic-buildimage and how can PR checker use it.

**Comment by vmittal-msft** [comment]
@ysmanman it is failing for coverage, Can you please check ?

**Comment by gechiang** [comment]
@vmittal-msft please share MSFT ADO number for this PR

**Comment by ysmanman** [comment]
> @ysmanman it is failing for coverage, Can you please check ?

Hi @vmittal-msft ,  I was trying to fix the test coverage, and then realized  a few issues/questions:

- Existing swss tests provides no test coverage for `PfcWdAclHandler`.  `PfcWdAclHandler` is only used for a few platforms: BFN or BRCM with DlrInit being disabled, e.g., DNX. I don't think existing tests cover any of these platforms.  This is the reason why code coverage check failed on my PR. 
-  According to my testing,  I didn't see PFCWD egress ACL table was removed after restoring from storm with existing `PfcWdAclHandler` implementation (without my change). Is this expected behavior or a bug?

**Comment by vmittal-msft** [comment]
> @vmittal-msft please share MSFT ADO number for this PR

ADO # 28523510

**Comment by prsunny** [comment]
Can we do mocktest coverage?

**Comment by vmittal-msft** [comment]
@bingwang-ms please help review. we need this for chassis project asap to relase image hence the priority,
@ysmanman please re-base.

**Comment by ysmanman** [comment]
> @bingwang-ms please help review. we need this for chassis project asap to release image hence the priority, @ysmanman please re-base.

Hi @vmittal-msft  PR was successfully rebased.

File: orchagent/aclorch.cpp
Comment (bingwang-ms) [inline orchagent/aclorch.cpp:880]: Return false here?

File: orchagent/aclorch.cpp
Comment (bingwang-ms) [inline orchagent/aclorch.cpp:4015]: This function is called after `createBindAclTable`. If the table has to be bound to switch, will that cause problem?

**Comment by bingwang-ms** [comment]
@kperumalbfn Can you please help review the PFCWD change? Thanks

**Comment by prsunny** [comment]
@kperumalbfn , kindly review. @ysmanman , do you've plans to fix coverage?

File: orchagent/aclorch.cpp
Comment (kperumalbfn) [inline orchagent/aclorch.cpp:5193]: Same as bind aclTable, please use ACL Group instead of ACLTable and switchorch already has the code.

**Comment by vmittal-msft** [comment]
@ysmanman please check comments and re-base.

**Comment by ysmanman** [comment]
> @kperumalbfn , kindly review. @ysmanman , do you've plans to fix coverage?

Hi @prsunny , I chatted with @vmittal-msft yesterday about the code coverage issue. The missing code coverage was due to  existing code not providing any code coverage for `PfcWdAclHandler`. So this is a test gap in current code. My PR unfortunately hit the test gap in coverage check. We need to fix the gap here. @vmittal-msft was suggesting to merge the PR first since the PR is required by 202405 and 202205 release, and then fixed coverage separately.

File: orchagent/aclorch.cpp
Comment (ysmanman) [reply orchagent/aclorch.cpp:4015]: > Instead of binding ACL table directly to switch, can we use ACL Table Group? If we use ACL Table, we could bind only one ACL to the switch bind point. ACL Table groups are used for port binding.

This was discussed in a previous meeting with both MSFT and BRCM. We understood the limitation, and decided to deferred this change later when we encounter the actual use case, i.e.,  we have multiple egress tables that need to bind to switch. @vmittal-msft

File: orchagent/aclorch.cpp
Comment (ysmanman) [reply orchagent/aclorch.cpp:880]: Fixed.

File: orchagent/aclorch.cpp
Comment (ysmanman) [reply orchagent/aclorch.cpp:4016]: We bind egress acl table to switch only for PFCWD table on DNX platform. This is fine because, on DNX,  PFCWD table does not bind port anymore. So `createBindAclTable` will only create PFCWD table on DNX, without binding the table to any port.

**Comment by prsunny** [comment]
> > @kperumalbfn , kindly review. @ysmanman , do you've plans to fix coverage?
> 
> Hi @prsunny , I chatted with @vmittal-msft yesterday about the code coverage issue. The missing code coverage was due to existing code not providing any code coverage for `PfcWdAclHandler`. So this is a test gap in current code. My PR unfortunately hit the test gap in coverage check. We need to fix the gap here. @vmittal-msft was suggesting to merge the PR first since the PR is required by 202405 and 202205 release, and then fixed coverage separately.

ok, we can take that separately. waiting for reviewers to sign-off

File: orchagent/aclorch.cpp
Comment (kperumalbfn) [inline orchagent/aclorch.cpp:5173]: Will be good to move this binding logic and invoking switch SAI API to switchOrchagent.

File: orchagent/aclorch.cpp
Comment (vmittal-msft) [reply orchagent/aclorch.cpp:4016]: Yes. This is was discussed in BRCM meeting and decided to bind acl table instead of group since there no current use case for this. If need arise in future, we can enhance this and support.

File: orchagent/aclorch.cpp
Comment (vmittal-msft) [reply orchagent/aclorch.cpp:5194]: Same as above.

File: orchagent/aclorch.cpp
Comment (vmittal-msft) [reply orchagent/aclorch.cpp:5173]: Since there are other functions in the same file which were using switch API to get information, so it was added here. I believe it should be ok.

Comment (kperumalbfn) [reply orchagent/aclorch.cpp:5173]: Please move the ACLTable binding to switch APIs to switchorch. We already have AclGroup switch bind in switchOrch.

Comment (ysmanman) [reply orchagent/aclorch.cpp]: Hi @kperumalbfn, moved `bindEgrAclTableToSwitch` and `unbindEgrAclTableFromSwitch` to switchorch.

File: orchagent/switchorch.cpp
Comment (kperumalbfn) [inline orchagent/switchorch.cpp:1520]: We could pass the aclStage and make this method common for both ingress and egress ACL.

File: orchagent/switchorch.cpp
Comment (kperumalbfn) [inline orchagent/switchorch.cpp:1540]: Same as above

**Review by kperumalbfn — COMMENTED** [review]
LGTM

**Review by kperumalbfn — APPROVED** [review]

File: orchagent/switchorch.cpp
Comment (ysmanman) [reply orchagent/switchorch.cpp]: Changed `bindEgrAclTableToSwitch` to `bindAclTableToSwitch` that takes aclStage as input.

File: orchagent/switchorch.cpp
Comment (ysmanman) [reply orchagent/switchorch.cpp]: Done.

**Comment by vmittal-msft** [comment]
@ysmanman  please re-base.

**Review by bingwang-ms — APPROVED** [review]

**Comment by bingwang-ms** [comment]
@ysmanman The new test [test_pfcwd_shared_egress_acl_table.py](https://github.com/sonic-net/sonic-swss/pull/3136/files#diff-b24af68ecf173f1d37778bb4630261e4107b64f12290996b7231aba10eec9d63) is consistently failing on `202405` branch. Can you please help check?
```

        """Run `polling_function` periodically using the specified `polling_config`.
    
        Args:
            polling_function: The function being polled. The function cannot take any arguments and
                must return a status which indicates if the function was succesful or not, as well as
                some return value.
            polling_config: The parameters to use to poll the polling function.
            failure_message: The message to print if the call times out. This will only take effect
                if the PollingConfig is set to strict.
    
        Returns:
            If the polling function succeeds, then this method will return True and the output of the
            polling function.
    
            If it does not succeed within the provided timeout, it will return False and whatever the
            output of the polling function was on the final attempt.
        """
        for _ in range(polling_config.iterations()):
            status, result = polling_function()
    
            if status:
                return (True, result)
    
            time.sleep(polling_config.polling_interval)
    
        if polling_config.strict:
            message = failure_message or f"Operation timed out after {polling_config.timeout} seconds with result {result}"
>           assert False, message
E           AssertionError: Operation timed out after 600 seconds with result None
```

**Pull request merged**
